Firestore/core/src/bundle/bundle_serializer.cc

```diff
@@ -98,31 +458,461 @@ BundleMetadata BundleSerializer::DecodeBundleMetadata(
+                                            const json& map_json) const {
+  if (!map_json.is_object() || !map_json.contains("fields")) {
+    context.Fail("mapValue is not a valid map");
+    return FieldValue();
+  }
+  const auto& fields = map_json.at("fields");
+  if (!fields.is_object()) {
+    context.Fail("mapValue's 'field' is not a valid map");
+    return FieldValue();
+  }
+
+  immutable::SortedMap<std::string, FieldValue> field_values;
+  for (auto it = fields.begin(); it != fields.end(); it++) {
+    field_values =
+        field_values.insert(it.key(), DecodeValue(context, it.value()));
+    if (!context.ok()) {
+      return FieldValue();
+    }
+  }
+
+  return FieldValue::FromMap(std::move(field_values));
+}
+
+FieldValue BundleSerializer::DecodeArrayValue(ReadContext& context,
+                                              const json& array_json) const {
+  if (!array_json.is_object() || !array_json.contains("values")) {
+    context.Fail("arrayValue is not a valid array object");
+    return FieldValue();
+  }
+  if (!array_json.at("values").is_array()) {
+    context.Fail("arrayValue is not a valid array");
+    return FieldValue();
+  }
+
+  const auto& values =
+      array_json.at("values").get_ref<const std::vector<json>&>();
+  std::vector<FieldValue> field_values;
+  for (const json& json_value : values) {
+    field_values.push_back(DecodeValue(context, json_value));
+    if (!context.ok()) {
+      return FieldValue();
+    }
+  }
+
+  return FieldValue::FromArray(std::move(field_values));
+}
+
+FieldValue BundleSerializer::DecodeReferenceValue(ReadContext& context,
+                                                  const json& ref_json) const {
+  auto ref_string = ToString(context, ref_json);
+  if (!context.ok()) {
+    return FieldValue();
+  }
+
+  // Calling into remote serializer with a separate context, to meet it's
+  // interface.
+  ReadContext separate_context;
+  ContextReader context_reader(separate_context);
+  auto result = rpc_serializer_.DecodeReference(&context_reader, ref_string);
+
+  // Copy status to our own context.
+  if (!context_reader.ok()) {
+    context.set_status(context_reader.status());
+    return FieldValue();
+  }
+
+  return result;
+}
+
+BundledDocumentMetadata BundleSerializer::DecodeDocumentMetadata(
+    util::ReadContext& context,
+    const std::string& document_metadata_string) const {
+  const json& document_metadata = Parse(document_metadata_string);
+
+  if (document_metadata.is_discarded()) {
+    context.Fail("Failed to parse string into json: " +
+                 document_metadata_string);
+    return BundledDocumentMetadata{};
+  }
+
+  if (!document_metadata.contains("name") ||
+      !document_metadata.contains("readTime")) {
+    context.Fail(
+        "One of bundledDocumentMetadata's 'name' or 'readTime' is missing");
+    return BundledDocumentMetadata{};
+  }
+
+  ResourcePath path = DecodeName(context, document_metadata.at("name"));
+  if (!context.ok()) {
```

Comment:
This is probably one of the rare cases where you do actually need an explicit check because `DocumentKey` asserts if its input is broken.